tsl/src/nodes/skip_scan/exec.c

```diff
@@ -0,0 +1,316 @@
+			break;
+
+		case SS_VALUES:
+			state->skip_key->sk_flags = 0;
+			break;
+
+		case SS_NULLS_LAST:
+		case SS_NULLS_FIRST:
+			state->skip_key->sk_flags = SK_ISNULL | SK_SEARCHNULL;
+			state->skip_key->sk_argument = 0;
+			break;
+
+		default:
+			break;
+	}
+
+	state->stage = new_stage;
+}
+
+static void
+skip_scan_update_key(SkipScanState *state, TupleTableSlot *slot)
+{
+	if (!state->prev_is_null && !state->distinct_by_val)
+		pfree(DatumGetPointer(state->prev_datum));
+
+	MemoryContext old_ctx = MemoryContextSwitchTo(state->ctx);
+	state->prev_datum = slot_getattr(slot, state->distinct_col_attnum, &state->prev_is_null);
+	if (state->prev_is_null)
+	{
+		state->skip_key->sk_flags = SK_ISNULL;
+		state->skip_key->sk_argument = 0;
+	}
+	else
+	{
+		state->prev_datum =
+			datumCopy(state->prev_datum, state->distinct_by_val, state->distinct_typ_len);
+		state->skip_key->sk_argument = state->prev_datum;
+	}
+
+	MemoryContextSwitchTo(old_ctx);
+
+	/* we need to do a rescan whenever we modify the ScanKey */
+	skip_scan_rescan_index(state);
+}
+
+static TupleTableSlot *
+skip_scan_exec(CustomScanState *node)
+{
+	SkipScanState *state = (SkipScanState *) node;
+	TupleTableSlot *result;
+
+	switch (state->stage)
+	{
+		case SS_NULLS_FIRST:
+			result = state->idx->ps.ExecProcNode(&state->idx->ps);
+
+			/*
+			 * if we found a NULL value we return it, otherwise
+			 * we restart the scan looking for non-NULL
+			 */
+			skip_scan_switch_stage(state, SS_NOT_NULL);
+			skip_scan_rescan_index(state);
+			if (!TupIsNull(result))
+				return result;
+			else
+				return skip_scan_exec(node);
+
+			break;
+
+		case SS_NOT_NULL:
+		case SS_VALUES:
+			result = state->idx->ps.ExecProcNode(&state->idx->ps);
+
+			if (!TupIsNull(result))
+			{
+				/*
+				 * if we found a tuple we update the skip scan key
+				 * and look for values greater than the value we just
+				 * found. If this is the first non-NULL value we
+				 * also switch stage to look for values greater than
+				 * that in subsequent calls.
+				 */
+				if (state->stage == SS_NOT_NULL)
+					skip_scan_switch_stage(state, SS_VALUES);
+
+				skip_scan_update_key(state, result);
+			}
+			else
+			{
+				/*
+				 * if there are no more values that satisfy
+				 * the skip constraint we are either done
+				 * for NULLS FIRST ordering or need to check
+				 * for NULLs if we have NULLS LAST ordering
+				 */
+				if (has_nulls_last(state))
+				{
+					skip_scan_switch_stage(state, SS_NULLS_LAST);
+					skip_scan_rescan_index(state);
+					return skip_scan_exec(node);
+				}
+				else
+				{
+					skip_scan_switch_stage(state, SS_END);
+				}
+			}
+			return result;
+			break;
+
+		case SS_NULLS_LAST:
+			result = state->idx->ps.ExecProcNode(&state->idx->ps);
+			skip_scan_switch_stage(state, SS_END);
+			return result;
+			break;
+
+		default:
+			return NULL;
+			break;
+	}
+}
+
+static void
+skip_scan_end(CustomScanState *node)
+{
+	SkipScanState *state = (SkipScanState *) node;
```

Comment:
Assert that state is SS_END here.

Reply:
that would not be correct, SS_END means the scan got exhausted but the scan can be ended before that by parent node e.g. a LIMIT node that has reached the limit